Set constants version via ScienceState classes

stargaser opened the pull request:
Fixes #6948 

Add ScienceState classes for `astropy.constants` versions, to allow a user to set constants versions before importing units or constants. Then when the user does `import astropy.units as u`, the user-specified versions of constants will initialize units.

EDIT: removed "work in progress" warnings
EDIT: renamed to use ScienceState instead of config files

mhvk commented:
Interesting! I certainly like that it is more obvious where the default is stored, and that by setting the constants right up front, the units are automatically OK.

stargaser commented:
@mhvk I think with the commit I just pushed modifying `constants/si.py`, it works. I need to put in several tests like this (after I modified my `astropy.cfg` to point to `codata2010` and `iau2012`):

```
In [1]: import astropy.units as u

In [2]: import astropy.constants as const

In [3]: (u.M_earth / const.M_earth).to(u.dimensionless_unscaled)
Out[3]: <Quantity 1.>
```

It will be critical to make sure the `constants.config` is modified before the first import of `units`.

mhvk commented:
That makes sense. It will be equally critical to document this clearly - while I do think it will be a feature that is rarely enough used that we can expect users to jump through some hoops, the hoops should be clear! And it is wonderful that this way it is possible to have consistent constants and units.

mhvk commented:
p.s.  Except...  I guess if we introduce new units based on new constants, we have a bit of a problem. Though it mostly means that some of the units code will have to check the constants configuration.

astrofrog commented on astropy/astropy.cfg:
```diff
@@ -60,6 +60,13 @@
+# physical_constants = 'codata2014'
+
+## Default version of astronomical constants
+# astronomical_constants = 'iau2015'
```
I think we shouldn't actually use config items here but ScienceState, as described here:

http://docs.astropy.org/en/stable/config/index.html#adding-new-configuration-items

stargaser replied:
I'm afraid I am not seeing how ScienceState can lead to consistent initialization of units, which is importing constants. It seems to me that config items for constants are needed before the first import of units.

stargaser replied:
Unless...as part of ScienceState, we do an `import lib.reload` of every core package that imports from constants?

stargaser replied:
In writing a test for the configuration items: I found  I can modify the `astropy.constants` configuration, then reload the `constants` package, and I get the desired version of constants. Unfortunately, I tried `importlib.reload(astropy.units)` and it would not pick up a modified `astropy.constants` configuration.

mhvk replied:
Reloading is quite flaky - I think one cannot easily change things that are already loaded inside a module.

mhvk commented:
@astrofrog - as noted by @stargaser, using a specific set of constants is easy only if it is the first thing you select, before importing all the units that rely on those constants. For this, a configuration item seems the easiest solution. Or can you see how to do it on-the-fly?

Note that this likely will be quite rare, so I think it is OK for there to be some hoops to jump through; in the end, for reproducibility, one would arguably be better of using an environment with fixed versions of all the software used.

stargaser commented:
The code coverage check has passed! A documentation section has been added too. I've removed the WIP annotations -- I think it is ready for @mhvk to take a closer look.

mhvk commented on astropy/constants/__init__.py:
```diff
@@ -23,12 +24,42 @@
+elif conf.physical_constants == 'codata2010':
+    from .astropyconst13 import iau2012 as iaudata
+else:
+    raise ValueError('Invalid configuration for physical constants: {}'
```
Copy & paste error: should be "astronomical" here, not physical (also in the `.format`).

stargaser replied:
Doh! Good catch! In the config test, I wasn't checking the prior version of astronomical constants, which I think would have caught this error. I've added a check.

mhvk commented on astropy/constants/tests/test_constants_config.py:
```diff
@@ -0,0 +1,43 @@
+        assert abs(h.si.value - 6.62606957e-34) < 1e-43
+        assert abs(h.cgs.value - 6.62606957e-27) < 1e-36
+
+        from astropy.constants.codata2014 import h as h_current
```
Nit: I'd call this `h_2014` since this won't stay "current" long!

stargaser commented:
For the larger question: I could add a config item "constants_version" which a user could set to 'astropyconst13' or 'astropyconst20'.  I think an exception should be raised if a user sets "constants_version" and "physical_constants" or "astronomical_constants".

mhvk commented:
Alternatively, it could be allowed to set both `physical_constants` and `astronomical_constants` to "astropy20" etc.

stargaser commented:
It gets tricky to implement "constants_version". When a user specifies only `constants_version='astropyconst13'`, the value of `conf.physical_constants` is still `'codata2014'` and the value of `conf.astronomical_constants` is still `'iau2015'`. We don't know if the user put either of those in the config file, or simply did not specify them.

mhvk commented:
So, how about my suggestion of allowing `physical_constants = 'astropy20'`? Then at least the answer should be unique...

stargaser commented:
Oh, got it! Don't add the extra config item, but allow the two items to take on values of `'astropyconst13'` and `'astropyconst20'` (to keep the naming consistent with the context manager), in addition to setting to `'codata2010'` and `'iau2012'`. Yes, that is a good idea. I will go implement that now.

mhvk commented on astropy/constants/__init__.py:
```diff
@@ -81,17 +118,23 @@ def set_enabled_constants(modname):
-            _utils._set_c(codata2014, iau2015, module,
+            _utils._set_c(codata, iaudata, module,
                           not_in_module_only=False, set_class=True)
+            del codata_context, iaudata_context
```
Is this necessary?  Wouldn't it happen automatically when execution reaches the end of this?

mhvk commented:
@stargaser - the failure on circle-ci looks related to what you have ...

stargaser commented:
Same failure on the doctest for the context manager. There's a very similar test in the module which passes. Unfortunately, I'm having trouble running doctests in my development environment -- `python setup.py test` stops with "path too long" error on my Macbook. I'm at a loss right now for how to fix this.

pllim commented:
There is still one failure on 32-bit
```
______________________ [doctest] docs/constants/index.rst ______________________
[gw0] linux -- Python 3.7.3 /opt/python/cp37-cp37m/bin/python
123 
124     >>> from astropy import constants as const
125     >>> with const.set_enabled_constants('astropyconst13'):
126     ...     print(const.h)
127       Name   = Planck constant
128       Value  = 6.62606957e-34
129       Uncertainty  = 2.9e-41
130       Unit  = J s
131       Reference = CODATA 2010
132     >>> print(const.h)
Differences (unified diff with -expected +actual):
    @@ -1,5 +1,5 @@
       Name   = Planck constant
    -  Value  = 6.62607004e-34
    -  Uncertainty  = 8.1e-42
    +  Value  = 6.62606957e-34
    +  Uncertainty  = 2.9e-41
       Unit  = J s
    -  Reference = CODATA 2014
    +  Reference = CODATA 2010
```

stargaser commented:
Working on it...Looks like I have to debug this in a 32-bit container. I set up a new development environment on a 64-bit Linux machine and all the tests passed with Python 3.7.3.

stargaser commented:
I am running the same container as CircleCI 32-bit and I have installed everything the same. All the tests pass for me. 

I am seeing a problem on all platforms in which  `import astropy.constants as const; print(const.h)` raises ValueErrors with "J is not a valid unit". If I first import astropy.units then everything is fine. I'll try a few more things but I am despairing of getting it all sorted today.

mhvk commented:
@stargaser - I'll try to have a look, that error is very strange

stargaser commented on astropy/constants/__init__.py:
```diff
@@ -23,12 +24,47 @@
 except ImportError:
```
I wonder if I have broken this initial import. In an interactive session, I have to `import astropy.units` before using a constant.

mhvk replied:
Is this still an issue?

stargaser replied:
Oh, this was something you had fixed by introducing `config.py`. Not an issue anymore. In fact the `try...except` does not appear to be needed.

mhvk commented:
Adding another "trial" import of units right before importing `si` solves it, but somehow I cannot just remove the one on top...

mhvk commented:
@stargaser - I think I have a solution: do the configuration in a separate submodule and import from there in `si` and `cgs`; I'll make a PR to your PR momentarily (or, if you are OK, can just push to your branch).

stargaser commented:
@mhvk -- that's great news! It's fine for you to just push to my branch. I'll add you.

mhvk commented:
@stargaser - well, I may have spoken too quickly. All the configuration tests now seem to fail...

mhvk commented:
OK, now things are OK again...

mhvk commented:
Now pushed to your branch. It solves the `print()` issue - hopefully also the 32 bit one!

mhvk commented:
@stargaser - as long as you allow pushes to your branch (which is the default for astropy), no need to invite me...

stargaser commented:
oh no! I saw the 32-bit build fail and I was trying to restart it -- and it restarted all. Apologies!

bsipocz commented:
> oh no! I saw the 32-bit build fail and I was trying to restart it -- and it restarted all. Apologies!

no worries. To restart only one job you need to select it from the dropdown.

bsipocz commented:
@mhvk - could you use the `merge-when-ci-passes` label when this is ready to go? I'll be online until quite late tonight so will be able to merge stuff.

astrofrog commented:
I’m still concerned that we shouldn’t have settings in the config file that affect the scientific results of a script (and that was the motivation for ScienceState). Is there no way around that in this case?

mhvk commented:
@astrofrog - The trouble is that we need to do something before we load units. In principle, I guess `ScienceState` could be used, if loaded before anything else, and we somehow check that if units is already loaded, it just fails.

p.s. I do not feel this is so hugely pressing that we could not postpone to 4.0; better to have a system we all are happy with! Though I think independently of anything else, we should figure out why the 32-bit build is not working!

stargaser commented:
32-bit failed again, with what looks like a problem with astropy-helpers. I tested the branch in a 32-bit container on my laptop with the same commands and it worked.

I've started a rerun of the job with ssh enabled -- not sure if I'll be online when it finishes but I thought I'd get it in the queue.

stargaser commented:
I'm okay with pushing this off to 4.0 -- maybe we can figure out better ways to do the imports to make `ScienceState` work.

astrofrog commented:
I think the 32-bit failure can be ignored.

What if the ScienceState lived at the root of the package, in astropy/__init__.py?

stargaser commented:
@astrofrog -- great suggestion! I think putting `ScienceState` in the Astropy root has a good chance of giving us what we want. I think we need something outside both constants and units, because once we fully import one or the other, it is too late to make units consistent with non-default constants.  The configuration system does its thing before those imports, and that is why we were able to get the configuration file to work.

My new goals for this PR are to have a solution using `ScienceState` ready for review close to World Metrology Day on 2019 May 20th, when the new SI takes effect and I am expecting renewed interest in the constants package.

eteq commented:
@stargaser - just wanted to check here: we've been waiting on the release to get the new values, but any progress here?  I.e., is there any chance of "fast-tracking" this one for 3.2, or should we continue to aim this for 4.0?

stargaser commented:
I'm afraid I haven't found time to push this forward lately. I've been assuming this feature is too big to put in the late stages of the 3.2 release cycle.

stargaser commented:
Starting work on putting a ScienceState class for physical constants into `astropy/__init__.py`...

stargaser commented:
I'm having trouble writing tests, because `units` and `constants` have already been imported when pytest gets to my test file. Would @mhvk or @astrofrog have suggestions for testing?

stargaser commented:
The functionality is there. I can do
```
>>> import astropy
>>> astropy.physical_constants.set('codata2010')
<ScienceState physical_constants: 'codata2010'>
>>> astropy.astronomical_constants.set('iau2012')
<ScienceState astronomical_constants: 'iau2012'>
>>> import astropy.units as u
>>> import astropy.constants as const
>>> (const.M_sun / u.M_sun).to(u.dimensionless_unscaled) - 1
<Quantity 0.>
>>> const.M_sun
<<class 'astropy.constants.iau2012.IAU2012'> name='Solar mass' value=1.9891e+30 uncertainty=5e+25 unit='kg' reference="Allen's Astrophysical Quantities 4th Ed.">
```

I can't seem to write that in a working test, because `units` and `constants` are already imported when pytest gets to my test file. Accordingly, the tests are simple.

mhvk commented on astropy/__init__.py:
```diff
@@ -157,6 +157,69 @@ class Conf(_config.ConfigNamespace):
+        if value not in cls._versions:
+            raise ValueError('Must be one of {}'
+                             .format(cls._versions.keys()))
+        return cls._versions.get(value)
```
Total nitpick, but I'd find `cls._versions[value]` clearer.

mhvk commented on astropy/constants/__init__.py:
```diff
@@ -68,10 +74,10 @@ def set_enabled_constants(modname):
     if modname == 'astropyconst13':
-        from .astropyconst13 import codata2010 as codata
-        from .astropyconst13 import iau2012 as iaudata
+        from .astropyconst13 import codata2010 as codata_context
```
In principle, one could try importing from the given module name. In that case, inside those modules `codata` and `iaudata` have to be defined.

But fine to do that later (maybe leave a comment?)

stargaser replied:
I think this is a good idea. Should we only have `codata` and `iaudata` defined inside `astropyconst13` and `astropyconst20`? 

We can also simplify the contents of `config.py` to:
```
import importlib
import astropy

phys_version = astropy.physical_constants.get()
astro_version = astropy.astronomical_constants.get()

codata = importlib.import_module('.constants.' + phys_version, 'astropy')
iaudata = importlib.import_module('.constants.' + astro_version, 'astropy')
```

mhvk replied:
And here I thought it would be just a small item - this now looks like it makes things considerably easier!

pllim commented on astropy/constants/config.py:
```diff
@@ -0,0 +1,33 @@
+elif ((physical_constants.get() == 'codata2010') or
+        (physical_constants.get() == 'astropyconst13')):
+    from .astropyconst13 import codata2010 as codata  # noqa
+else:
```
If this will never happen, should exclude it from coverage.
```suggestion
else:  # pragma: no cover
```

stargaser replied:
I've revamped `config.py`. The validation happens in the ScienceState classes so doesn't need a test here.

pllim commented on astropy/constants/config.py:
```diff
@@ -0,0 +1,33 @@
+elif ((astronomical_constants.get() == 'iau2012') or
+        (astronomical_constants.get() == 'astropyconst13')):
+    from .astropyconst13 import iau2012 as iaudata  # noqa
+else:
```
```suggestion
else:  # pragma: no cover
```

stargaser replied:
The modules has been revamped.

pllim commented on astropy/constants/tests/test_sciencestate.py:
```diff
@@ -0,0 +1,24 @@
+def test_version_match():
+    from astropy import physical_constants, astronomical_constants
+    pversion = physical_constants.get()
+    import astropy.constants as const
```
If there is a reason this import have to happen after `.get()`, it should be documented, lest in the future, someone else moves it back to the top.

stargaser replied:
No reason for it now. Earlier, I was experimenting with deleting `units` and `constants` from `sys.modules`, but that caused units tests to fail.

pllim commented:
CircleCI 32-bit is weird. It appears to be running under your own CircleCI instance, not Astropy's. I think @nden ran into this before and the problem went away when she removed her own CircleCI instance (or something like that).

nden commented:
Yes, the only way I could "correct" this is to close astropy testing on my Circleci account.

stargaser commented on astropy/__init__.py:
```diff
@@ -157,6 +157,69 @@ class Conf(_config.ConfigNamespace):
+    def validate(cls, value):
+        if value not in cls._versions:
+            raise ValueError('Must be one of {}'
+                             .format(cls._versions.keys()))
```
I think a `list()` is needed. I hadn't thought about sorting -- excellent point! Astropy 4.0 will be 3.6 and above, right?

pllim replied:
According to https://github.com/astropy/astropy-APEs/blob/master/APE10.rst , yes.

stargaser replied:
I've implemented the `list()` but not the sort. The exception messages list the options in order of insertion:
```
ValueError: Must be one of ['codata2018', 'codata2014', 'codata2010', 'astropyconst40', 'astropyconst20', 'astropyconst13']

ValueError: Must be one of ['iau2015', 'iau2012', 'astropyconst40', 'astropyconst20', 'astropyconst13']
```

stargaser commented on astropy/__init__.py:
```diff
@@ -157,6 +157,69 @@ class Conf(_config.ConfigNamespace):
+        return ctx
+
+
+class astronomical_constants(physical_constants):
```
Yes, this would make more sense.

mhvk commented:
@pllim is right about the documentation: Probably should have something in `units` as well.

stargaser commented on astropy/constants/__init__.py:
```diff
@@ -37,9 +41,11 @@
-              not_in_module_only=True, doclines=_lines, set_class=True)
+# Catch warnings about "already has a definition in the None system"
+with warnings.catch_warnings():
+    warnings.simplefilter('ignore')
```
Great recommendation! I'll filter on a regex of the warning message.

stargaser commented:
Thanks @pllim and @mhvk for all the review comments! I believe I have addressed the code comments. The documentation is still to do. I've fixed the CircleCI problem. Tests are all passing now!

After the docs, I really need to squash commits. I'll save a copy of this branch first.

bsipocz commented on docs/units/constants_versions.rst:
```diff
@@ -11,15 +11,15 @@ Setting the prior versions at the start of a Python session will allow
 
 >>> import astropy
->>> astropy.physical_constants.set('codata2010')
+>>> astropy.physical_constants.set('codata2010')  # doctest: +SKIP
```
why skipping the doctesting?

stargaser replied:
It raises RuntimeError, because units or constants have already been imported.

bsipocz replied:
I think it would still be preferable to show a working example, let me dig up how to reset the pytest namespace.

mhvk replied:
If one can reset the namespace, that might solve the testing problem more generally too...

bsipocz replied:
now this is embarrassing as I remembered to this to be something easy to do, but then now can't find how to do it. So I would suggest that we open a separate issue for this for 4.0 to make sure this will be attended with a solution by the time this is released to users.

mhvk replied:
Sounds good. @stargaser - could you raise that issue?

bsipocz replied:
I'll open an issue in a moment.

mhvk commented on astropy/constants/__init__.py:
```diff
@@ -68,10 +70,10 @@ def set_enabled_constants(modname):
     from . import utils as _utils
 
-    # NOTE: Update this when default changes.
+    # NOTE: Update this whenever versions are added.
```
We could avoid having to remember to update by doing something along the lines of what you do in `config.py`, say
```
try:
    codata = importlib.importmodule('.' + modname + '.codata', 'astropy')
    iaudata = importlib.importmodule('.' + modname + '.iaudata', 'astropy')
except ImportError as exc:
    exc.args += ('Context manager does not currently handle {}'.format(modname),)
    raise
```

stargaser replied:
Great idea! I've implemented it, slightly differently since it e.g. `astropy.constants.astropyconst13.codata` is not recognized as a module. But `astropy.constants.astropyconst13` is.

bsipocz commented:
I'm puzzled:
```
======================================================= FAILURES =======================================================
_____________________________________ [doctest] docs/units/constants_versions.rst ______________________________________
018 >>> import astropy
019 >>> astropy.physical_constants.set('codata2010')
020 <ScienceState physical_constants: 'codata2010'>
021 >>> astropy.astronomical_constants.set('iau2012')
022 <ScienceState astronomical_constants: 'iau2012'>
023 >>> import astropy.units as u
024 >>> import astropy.constants as const
025 >>> (const.M_sun / u.M_sun).to(u.dimensionless_unscaled) - 1   # doctest: +FLOAT_CMP
026 <Quantity 0.>
027 >>> print(const.M_sun)
Differences (unified diff with -expected +actual):
    @@ -1,5 +1,5 @@
       Name   = Solar mass
    -  Value  = 1.9891e+30
    -  Uncertainty  = 5e+25
    +  Value  = 1.9884754153381438e+30
    +  Uncertainty  = 9.236140093538353e+25
       Unit  = kg
    -  Reference = Allen's Astrophysical Quantities 4th Ed.
    +  Reference = IAU 2015 Resolution B 3 + CODATA 2014

/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-78uso7fm/lib/python3.7/site-packages/docs/units/constants_versions.rst:27: DocTestFailure
```

while if I copy paste the code snippet into ipython I get the same output (from Allen's Astrophysical Quantities 4th Ed.).

bsipocz commented:
@stargaser - could you please rebase?

mhvk commented:
:fireworks:

bsipocz commented:
Could you comment on the failure I'm seeing in https://github.com/astropy/astropy/pull/8517#issuecomment-497876801?

stargaser commented:
> Could you comment on the failure I'm seeing in [#8517 (comment)](https://github.com/astropy/astropy/pull/8517#issuecomment-497876801)?

 I see the same behavior -- works in a new Python or IPython session, and not in doctest or pytest. #8781 is for these discrepancies so I recommend to discuss there.
